### DISCUSSION TITLE
Handling of reportUnknownArgumentType  with Any?

### DISCUSSION THREAD

**Discussion by sherbang** [microsoft/pyright#6939] [General]
With `reportUnknownArgumentType` on, should Any types be recognized as unknown?

In this example, I expected sub_test to warn me that `p.any_val` is an unknown type.  It seems that there isn't any way to specify the type of p where I know the first generic type, but don't know the second.

```python
@dataclass
class Data[T_str, T_any]:
    str_val: T_str
    any_val: T_any
    


def sub_test(p: str):
    ...


def test(p: Data[str, Any]):
    assert_type(p.any_val, str)
    sub_test(p.any_val)
```

I've seen some other answers here and in the documentation that talk about the difference between explicit and implicit unknowns.  I appreciate the desire to give the developer more control here.  Perhaps a step further in this direction would be a `reportAnyArgumentType` setting?  I'm [not the only person](https://github.com/microsoft/pyright/discussions/4867) who would benefit from such an option.

**Comment by erictraut** [comment]
Pyright is working correctly in this case.

The type of `p.any_val` is `Any` because you have explicitly specialized the class with `Any`. An `Unknown` is an "implicit" `Any`, one where you have not explicitly specified a type so the type checker needs to assume `Any`.

If you were to omit the type arguments for `p: Data`, then the type of the second type argument would be `Unknown`.

**Reply by sherbang** [reply]
Yes, I understand that.  In my use-case, I need to specify the first type, but the second will be unknown in this block.

**Reply by erictraut** [reply]
You could use `Data[str]`. Pyright will fill in the second type argument with `Unknown`.

Keep in mind that `Unknown` is not an official part of the Python type system, so there's no way to "spell" it directly. It's simply pyright's way to note that you probably omitted something in your code.

Why do you want to generate an `Unknown` here? It's generally undesirable to do so.

**Reply by sherbang** [reply]
I started with something like this:
```python
from dataclasses import dataclass


@dataclass(frozen=True)
class Data[T]:
    id_: str
    val: T


class DataHandler[T]:
    def get_by_id(self, id: str) -> T | None:
        ...

    def _save(self, d: T) -> None:
        ...

    def update(self, d: Data[T]) -> None:
        existing = self.get_by_id(d.id_)
        if existing:
            ...
        else:
            self._save(d.val)
```

I started updating my code to use `NewType` for my IDs so that I don't cross IDs somewhere.  This resulted in `id_` being made generic in the `Data` class. So I made the change and started going through pyright errors to find what I broke. I updated this code to:
```python
from dataclasses import dataclass
from typing import Any, NewType


@dataclass(frozen=True)
class Data[ValT, IdT]:
    id_: IdT
    val: ValT


DataHandlerIDType = NewType("DataHandlerIDType", str)


class DataHandler[T]:
    def get_by_id(self, id: DataHandlerIDType) -> T | None:
        ...

    def _save(self, d: T) -> None:
        ...

    def update(self, d: Data[T, Any]) -> None:
        existing = self.get_by_id(d.id_)
        if existing:
            ...
        else:
            self._save(d.val)
```

With the `get_by_id()` call buried in the `update()` method, I didn't realize it existed at all.  I added `Any` to the `Data` type definition because I don't want to type-ignore the whole `update()` definition with `pyright: ignore[reportGeneralTypeIssues]`.

The type of `d.id_` is sometimes incompatible with `DataHandler.get_by_id()`.  Since `Data` is frozen I opted to do my id conversion separately and add it as a separate parameter rather than make another copy of everything within `Data`. So now my code looks like:
```python
from dataclasses import dataclass
from typing import Any, NewType


@dataclass(frozen=True)
class Data[ValT, IdT]:
    id_: IdT
    val: ValT


DataHandlerIDType = NewType("DataHandlerIDType", str)


class DataHandler[T]:
    def get_by_id(self, id: DataHandlerIDType) -> T | None:
        ...

    def _save(self, d: T) -> None:
        ...

    def update(self, d: Data[T, Any], id_: DataHandlerIDType) -> None:
        existing = self.get_by_id(id_)
        if existing:
            ...
        else:
            self._save(d.val)
```

...and I find myself having to choose between:
- updating the definition of `DataHandler`, requiring every user of it to specify the type of a value that doesn't get referenced
- Removing `Any` from `update()` and annotating it with `pyright: ignore[reportGeneralTypeIssues]` (masking additional future type errors on that line)
- Leaving `Any` in, and not getting a error if someone starts using `d.id_` in this method in the future

**Reply by erictraut** [reply]
Have you considered using `d: Data[T, DataHandlerIDType]` rather than `d: Data[T, Any]`?

**Reply by sherbang** [reply]
Yes, but that type doesn't match the incoming data.

I did consider requiring callers to conform the id type, but that would require callers to construct a new Data object with the new id.  With a dataclass that's not too bad (`dataclasses.replace()`), but my actual code is using pydantic so it would probably involve dumping the model to a dict, and then to a new model.  Instead, I added the new `id_` parameter to pass the id separately.

---

Another similar example is using the request object in litestar (and it looks like this is also true in the underlying starlette library):
```python
from litestar import get, Request

@get('/')
def root(request:Request[Any,Any,Any]):
    ...
```

The request object has 3 optional type parameters that define the types of the user, auth, and state data within the request.  Often I'll need access to the request, and not any of these pieces of data, or perhaps just one of them.

Should I be annotating the types of all 3 of these parameters every time I use request?

What if I'm writing a reusable Litestar module that provides some endpoints?  Different applications that reuse my code will have different types for these generics.  I'd need to provide a way for users of my module to specify the types of these variables, this would probably require a major refactor of my module, for variables that I don't even use.

Isn't this the point of `Any`?  It's to specify explicitly that I don't know what type this variable will have.  Just because I don't know the type though, doesn't mean that I want to be warned if I later try to use it as a `str`.  I'd expect a variable of `Any` type would only be able to be passed to a function that accepts `Any`.

**Comment by CleaverClamChowder** [comment]
Thank you

BigDawg83

**Comment by sherbang** [comment]
I found the answer that I'm looking for.

I should be defining `test` as:
```python
def test(p: Data[str, object]):
```

My replacing `Any` with `object` it still allows all types to be passed in, but shows where I'm using the wrong type within my function.
